zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server.auth.znode.groupacl;
+
+import java.security.cert.X509Certificate;
+import java.util.HashSet;
+import java.util.Set;
+import javax.net.ssl.X509KeyManager;
+import javax.net.ssl.X509TrustManager;
+import javax.security.auth.x500.X500Principal;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.common.ZKConfig;
+import org.apache.zookeeper.data.Id;
+import org.apache.zookeeper.server.ServerCnxn;
+import org.apache.zookeeper.server.auth.ServerAuthenticationProvider;
+import org.apache.zookeeper.server.auth.X509AuthenticationUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A ServerAuthenticationProvider implementation that does both authentication and authorization for protecting znodes from unauthorized access.
+ * Znodes are grouped into domains according to their ownership, and clients are granted access permission to domains.
+ * Authentication mechanism is same as in X509AuthenticationProvider.
+ * Authorization is done by checking with clients' URI (uniform resource identifier) in ACL metadata for matched domains.
+ * This class is meant to support the following use patterns:
+ *  1.Multiple creators, one reader - ZNodes, created by multiple identities from multiple domains,
+ *    need to be read by an identity from a different domain.
+ *  2.Multiple creators, multiple readers - ZNodes, created by multiple identities from multiple domains,
+ *    need to be read by multiple identities from multiple domains.
+ *  3.Multiple creators, one all-permission accessor - ZNodes, created by multiple identities from multiple domains,
```

Comment:
let's re-work this. good try, but confusing. here's where it's confusing:

`multiple creators, one reader - ZNodes, ` -> multiple creators? what does this mean? multiple clients belonging to one domain? multiple domains? 

and why are they all multiple domains?